### PR TITLE
CMFD implementation for MOC solver 

### PR DESCRIPTION
**Pull request by gaiarj**
This PR implements Coarse Mesh Finite Difference (CMFD) acceleration into the MOC solver for both isotropic and anisotropic scattering. It is also implemented for both fixed source and k-eigenvalue problems. The implementation is contained within the CMFD class but also includes modifications to the MOCDriver, Track, and Segment classes in order to facilitate the tallying of currents on CMFD cell surfaces which is necessary for the method to work.  

The basic functionality of the CMFD implementation is as follows:
While the MOC problem has not converged:
Complete a MOC sweep
->Enter CMFD:
  1. Normalize tallied currents by the surface lengths
  2. Homogenize cross sections and flux spectrum from the MOC solver
  3. Create the loss and source matrices to solve the diffusion problem
  4. Solve the diffusion problem (directly for fixed source, power iteration for k-eigenvalue)
  5. Update MOC scalar fluxes and boundary angular fluxes using the CMFD results 
 
->Exit CMFD:
Set MOC Keff equal to CMFD Keff 

Test cases have been included in tests/cmfd for both anisotropic/isotropic scattering, all 3 possible boundary conditions, and both fixed source and k-eigenvalue problems. 

As an example of the acceleration this feature provides, the C5G7 example in examples/c5g7.py has been updated to use CMFD, with the default settings and three group energy condensation. It now takes only 28 MOC iterations to converge to 1E-5 flux and keff difference, compared to 899 MOC iterations before

### CODE REVIEW

File: src/scarabee/_scarabee/include/moc/cmfd.hpp [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+  void create_loss_matrix(const MOCDriver& moc);
+  void create_source_matrix();
+  void power_iteration(double keff);
+  void fixed_source_solve();
+  void update_moc_fluxes(MOCDriver& moc);
+  void normalize_currents();
+  void compute_homogenized_xs_and_flux(const MOCDriver& moc);
+  void check_neutron_balance(const std::size_t i, const std::size_t j,
+                             std::size_t g, const double keff) const;
+};
```
Comment (HunterBelanger) [inline src/scarabee/_scarabee/include/moc/cmfd.hpp:209]: At the end of all the private methods we should have a serialization method like can be found in other parts of the code. This allows us to save the entire object to disk and reload it if we want.

File: src/scarabee/_scarabee/python/cmfd.cpp
Comment (HunterBelanger) [inline src/scarabee/_scarabee/python/cmfd.cpp:1]: A few notes for this file. First, try to make sure there is a blank line between binding methods. Second, make sure that the type hints correspond to the python type ! For example, I see many unsigned ints, but this type doesn't exist in Python, and is only an int.

File: src/scarabee/_scarabee/python/moc_driver.cpp [hunk 2]
```diff
@@ -102,6 +102,9 @@ void init_MOCDriver(py::module& m) {
                     "pass, a warning is issued, but the calculation continues. "
                     "Default value is False.")
 
+      .def_property("cmfd", &MOCDriver::cmfd, &MOCDriver::set_cmfd,
+                    "CMFD mesh for convergence acceleration.")
```
Comment (HunterBelanger) [inline src/scarabee/_scarabee/python/moc_driver.cpp:106]:
```suggestion
                    "CMFD object for convergence acceleration.")
```

File: src/scarabee/_scarabee/moc_driver.cpp [hunk 3]
```diff
@@ -104,6 +106,20 @@ void MOCDriver::set_keff_tolerance(double ktol) {
...
+void MOCDriver::set_cmfd(std::shared_ptr<CMFD> cmfd) {
+  if ((this->drawn())) {
+    angle_info_.clear();
+    tracks_.clear();
+    spdlog::warn(
+        "CMFD was set after track tracing. Must call generate_tracks again !");
+  }
+
+  // TODO Add check that MOCDriver width is same as CMFD width (to floating
+  // point tolerance)
```
Comment (HunterBelanger) [inline src/scarabee/_scarabee/moc_driver.cpp:118]: I thought you had implemented this already ?

File: src/scarabee/_scarabee/moc_driver.cpp [hunk 4]
```diff
@@ -333,7 +355,15 @@ void MOCDriver::solve_isotropic() {
...
-    if (mode_ == SimulationMode::Keff) {
+    if (mode_ == SimulationMode::Keff && !cmfd_) {
+      prev_keff = keff_;
+      keff_ = calc_keff(next_flux, flux_);
+      rel_diff_keff = std::abs(keff_ - prev_keff) / keff_;
+    }
+
+    // If MOC iterations are skipped compute Keff
+    // the normal way
+    if (mode_ == SimulationMode::Keff && cmfd_ && !cmfd_->solved()) {
```
Comment (HunterBelanger) [inline src/scarabee/_scarabee/moc_driver.cpp:366]: These two if statements do exactly the same thing. Please try to change the logic so that they are together instead of duplicating this code.

File: src/scarabee/_scarabee/moc_driver.cpp [hunk 5]
```diff
@@ -353,6 +383,16 @@ void MOCDriver::solve_isotropic() {
...
     flux_ = next_flux;
 
+    // Apply CMFD
+    if (cmfd_) {
+      cmfd_->solve(*this, prev_keff, iteration);
+      if (cmfd_->solved()) {
+        prev_keff = keff_;
+        keff_ = cmfd_->keff();
+        rel_diff_keff = std::abs(keff_ - prev_keff) / keff_;
+      }
```
Comment (HunterBelanger) [inline src/scarabee/_scarabee/moc_driver.cpp:393]: So this is only needed if running in eigenvalue mode.

File: src/scarabee/_scarabee/moc_driver.cpp [hunk 6]
```diff
@@ -453,9 +495,18 @@ void MOCDriver::solve_anisotropic() {
...
-    if (mode_ == SimulationMode::Keff) {
+    if (mode_ == SimulationMode::Keff && !cmfd_) {
+      prev_keff = keff_;
+      keff_ = calc_keff(next_flux, flux_);
+      rel_diff_keff = std::abs(keff_ - prev_keff) / keff_;
+    }
+
+    // If MOC iterations are skipped compute Keff
+    // the normal way
+    if (mode_ == SimulationMode::Keff && cmfd_ && !cmfd_->solved()) {
```
Comment (HunterBelanger) [inline src/scarabee/_scarabee/moc_driver.cpp:509]: As before, please combine these blocks of identical code.

File: src/scarabee/_scarabee/moc_driver.cpp [hunk 7]
```diff
@@ -479,6 +530,16 @@ void MOCDriver::solve_anisotropic() {
...
     flux_ = next_flux;
 
+    // Apply CMFD
+    if (cmfd_) {
+      cmfd_->solve(*this, prev_keff, iteration);
+      if (cmfd_->solved()) {
+        prev_keff = keff_;
+        keff_ = cmfd_->keff();
+        rel_diff_keff = std::abs(keff_ - prev_keff) / keff_;
+      }
```
Comment (HunterBelanger) [inline src/scarabee/_scarabee/moc_driver.cpp:540]: Only needed in k-eigenvalue.

File: src/scarabee/_scarabee/moc_driver.cpp [hunk 8]
```diff
@@ -607,14 +721,39 @@ void MOCDriver::sweep_anisotropic(xt::xtensor<double, 3>& sflux,
         const double tw = 4. * PI * track.wgt() *
                           track.width();  // Azimuthal weight * track width
 
+        // Get the azimuthal angle (phi) and its cosine for CMFD current
+        const Direction u_forw = track.dir();
+        const Direction u_back = -u_forw;
+
+        // Accumulate entry angular flux into CMFD current
+        // TODO : The entry angular flux should also be multiplied by the
+        // spherical harmonic?
```
Comment (HunterBelanger) [inline src/scarabee/_scarabee/moc_driver.cpp:730]: I think we resolved this, right ?

File: src/scarabee/_scarabee/moc_driver.cpp [hunk 9]
```diff
@@ -657,14 +805,36 @@ void MOCDriver::sweep_anisotropic(xt::xtensor<double, 3>& sflux,
         for (std::size_t pp = 0; pp < n_pol_angles_; pp++)
           angflux[pp] = track.exit_flux()(g, pp);
 
+        // Accumulate entry angular flux into CMFD current for backwards
+        // direction
+        // TODO : The entry angular flux should also be multiplied by the
+        // spherical harmonic?
```
Comment (HunterBelanger) [inline src/scarabee/_scarabee/moc_driver.cpp:811]: We resolved this, right ?

File: src/scarabee/_scarabee/cmfd.cpp [hunk 10]
```diff
@@ -134,8 +130,32 @@ CMFD::CMFD(const std::vector<double>& dx, const std::vector<double>& dy,
...
-  // Allocate the flux array
+  // Allocate cell volume array
+  volumes_.resize(nx_ * ny_);
+  for (std::size_t i = 0; i < nx_; i++) {
+    for (std::size_t j = 0; j < ny_; j++) {
+      // Store CMFD cell volume
+      const auto indx = tile_to_indx(i, j);
+      volumes_[indx] = dx_[i] * dy_[j];
+    }
+  }
```
Comment (HunterBelanger) [inline src/scarabee/_scarabee/cmfd.cpp:141]: Do we really need a volumes array ? I thought we had considered removing it ? Because we don't need it often and can easily just do dx_[i] * dy_[j].

File: src/scarabee/_scarabee/cmfd.cpp [hunk 11]
```diff
@@ -134,8 +130,32 @@ CMFD::CMFD(const std::vector<double>& dx, const std::vector<double>& dy,
...
+
+  // Set CMFD fluxes to 1
+  flux_cmfd_.resize(ng_ * nx_ * ny_);
+  flux_cmfd_.setOnes();
+
+  // Allocate flux update ratio array
+  update_ratios_.resize(ng_ * nx_ * ny_);
+  update_ratios_.setZero();
```
Comment (HunterBelanger) [inline src/scarabee/_scarabee/cmfd.cpp:149]: I would initialize this to ones, just so we don't end up with a zero flux everywhere on accident one day.

File: src/scarabee/_scarabee/cmfd.cpp [hunk 12]
```diff
@@ -187,14 +213,33 @@ std::optional<std::size_t> CMFD::get_surface(const Vector& r,
...
+    surface.crossing = CMFDSurfaceCrossing::Type::XN;
+  } else if (dy_p < SURFACE_COINCIDENT) {
+    surface.crossing = CMFDSurfaceCrossing::Type::YP;
+  } else if (dy_n < SURFACE_COINCIDENT) {
+    surface.crossing = CMFDSurfaceCrossing::Type::YN;
+  }
 
-  // If we get here, we aren't on a surface
-  return std::nullopt;
+  else {
```
Comment (HunterBelanger) [inline src/scarabee/_scarabee/cmfd.cpp:236]: Please move this up to be attached to the closing bracket of the previous else if statement.

File: src/scarabee/_scarabee/cmfd.cpp [hunk 13]
```diff
@@ -187,14 +213,33 @@ std::optional<std::size_t> CMFD::get_surface(const Vector& r,
...
+    surface.is_valid = false;
+    return surface;
+  }
+  surface.is_valid = true;
+
+  return surface;
```
Comment (HunterBelanger) [inline src/scarabee/_scarabee/cmfd.cpp:242]:
```suggestion
  }
  
  surface.is_valid = true;
  return surface;
```

File: src/scarabee/_scarabee/cmfd.cpp [hunk 14]
```diff
@@ -234,6 +290,23 @@ std::size_t CMFD::moc_to_cmfd_group(std::size_t g) const {
...
+  if (g > ng_) {
+    auto mssg = "Group index out of range.";
+    spdlog::error(mssg);
+    throw ScarabeeException(mssg);
+  }
+  if (i >= nx_ || j >= ny_) {
+    auto mssg = "Indexed cell does not exist.";
+    spdlog::error(mssg);
+    throw ScarabeeException(mssg);
+  }
```
Comment (HunterBelanger) [inline src/scarabee/_scarabee/cmfd.cpp:304]: Check the i and j separately so that you can say which index is out of range to the user.

File: src/scarabee/_scarabee/cmfd.cpp [hunk 15]
```diff
@@ -325,26 +456,764 @@ void CMFD::compute_homogenized_xs_and_flux(const MOCDriver& moc) {
...
+    auto mssg = "Tolerance for flux must be in the interval (0., 0.1).";
+    spdlog::error(mssg);
+    throw ScarabeeException(mssg);
+  }
+
+  if (ftol >= 0.1) {
+    auto mssg = "Tolerance for flux must be in the interval (0., 0.1).";
+    spdlog::error(mssg);
+    throw ScarabeeException(mssg);
+  }
```
Comment (HunterBelanger) [inline src/scarabee/_scarabee/cmfd.cpp:562]: You can put these in the same if statement.

File: src/scarabee/_scarabee/cmfd.cpp [hunk 16]
```diff
@@ -325,26 +456,764 @@ void CMFD::compute_homogenized_xs_and_flux(const MOCDriver& moc) {
...
+    auto mssg = "Tolerance for keff must be in the interval (0., 0.1).";
+    spdlog::error(mssg);
+    throw ScarabeeException(mssg);
+  }
+
+  if (ktol >= 0.1) {
+    auto mssg = "Tolerance for keff must be in the interval (0., 0.1).";
+    spdlog::error(mssg);
+    throw ScarabeeException(mssg);
+  }
```
Comment (HunterBelanger) [inline src/scarabee/_scarabee/cmfd.cpp:578]: These should go in the same if statement.

File: src/scarabee/_scarabee/cmfd.cpp [hunk 17]
```diff
@@ -325,26 +456,764 @@ void CMFD::compute_homogenized_xs_and_flux(const MOCDriver& moc) {
...
+    case CMFD::TileSurf::YN:
+      return surface_currents_(g, get_y_neg_surf(i, j));
+      break;
+  }
+
+  // NEVER GETS HERE
+  return 0.;
+}
+
+void CMFD::larsen_correction(double& D, const double dx,
```
Comment (HunterBelanger) [inline src/scarabee/_scarabee/cmfd.cpp:675]: Maybe add the reference for this paper too.

File: src/scarabee/_scarabee/cmfd.cpp [hunk 18]
```diff
@@ -325,26 +456,764 @@ void CMFD::compute_homogenized_xs_and_flux(const MOCDriver& moc) {
...
+  const double EtDx = Et_(g, i, j) * dx;
+
+  if (EtDx < 1.) {
+    return;
+  } else if (EtDx < 14 && EtDx >= 1) {
+    const double theta =
+        -5.542780E-02 + 8.740501E-02 * EtDx +
+        -2.152599E-02 * std::pow(EtDx, 2) + 3.145553E-03 * std::pow(EtDx, 3) +
+        -2.683648E-04 * std::pow(EtDx, 4) + 1.222516E-05 * std::pow(EtDx, 5) +
+        -2.284879E-07 * std::pow(EtDx, 6);
```
Comment (HunterBelanger) [inline src/scarabee/_scarabee/cmfd.cpp:719]: This could likely be a little faster if you replace the `std::pow` calls with computing your own powers, as you need all of them anyway up to 6.

File: src/scarabee/_scarabee/cmfd.cpp [hunk 19]
```diff
@@ -325,26 +456,764 @@ void CMFD::compute_homogenized_xs_and_flux(const MOCDriver& moc) {
...
+      if (ratio > 20.0 && moc_iteration_ > 1) {
+        flux_update_warning = true;
+      }
+      // Check that the update ratio is valid
+      if (ratio < 0.0) {
+        auto mssg =
+            "Negative CMFD flux update ratio. Try using Larsen correction";
+        spdlog::error(mssg);
+        throw ScarabeeException(mssg);
+      }
```
Comment (HunterBelanger) [inline src/scarabee/_scarabee/cmfd.cpp:1093]: Can we add another check for ration == 0.0 ?

File: src/scarabee/_scarabee/cmfd.cpp [hunk 20]
```diff
@@ -325,26 +456,764 @@ void CMFD::compute_homogenized_xs_and_flux(const MOCDriver& moc) {
...
+
+    /**
+    for (std::size_t i = 0; i < nx_; i++) {
+      for (std::size_t j = 0; j < ny_; j++) {
+        for (std::size_t g = 0; g < ng_; g++) {
+          this->check_neutron_balance(i, j, g, keff);
+        }
+      }
+    }
+    */
```
Comment (HunterBelanger) [inline src/scarabee/_scarabee/cmfd.cpp:1207]: IDK if we should make this optional or not ?

Comment (gaiarj) [reply src/scarabee/_scarabee/cmfd.cpp:1207]: I think it should be another option for the CMFD class, off by default. That is probably a good idea

File: src/scarabee/_scarabee/include/moc/cmfd.hpp [hunk 21]
```diff
@@ -0,0 +1,213 @@
...
+  double get_damping() const { return damping_; }
+
+  void set_flux_limiting(bool user_pref) { flux_limiting_ = user_pref; }
+  bool get_flux_limiting() const { return flux_limiting_; }
+
+  void set_larsen_correction(bool user_pref) { larsen_correction_ = user_pref; }
+  bool get_larsen_correction() const { return larsen_correction_; }
+
+  void set_od_cmfd(bool user_pref) { od_cmfd_ = user_pref; }
+  bool get_od_cmfd() const { return od_cmfd_; }
```
[the remark this answers is not in the document]
Comment (gaiarj) [reply src/scarabee/_scarabee/include/moc/cmfd.hpp:115]: I had it check in the solve method and error if both are set,  which isn't the best solution. It should be in the setter so the user doesn't waste time, I agree

File: src/scarabee/_scarabee/moc_driver.cpp (hunk 8 above)
Comment (gaiarj) [reply src/scarabee/_scarabee/moc_driver.cpp:730]: Yes, the comment should be removed

File: src/scarabee/_scarabee/cmfd.cpp (hunk 10 above)
Comment (gaiarj) [reply src/scarabee/_scarabee/cmfd.cpp:141]: We could do dx_[i] * dy_[j] at the moment it is needed, but that quantity is known at the construction and so can be computed just once. The increase in performance may be negligible though...

File: src/scarabee/_scarabee/moc_driver.cpp (hunk 4 above)
Comment (HunterBelanger) [reply src/scarabee/_scarabee/moc_driver.cpp:366]: So I believe the following logic will work to combine them into a single if statement:

```
if (mode_ == SimulationMode::Keff && 
    (cmfd_ == nullptr  || (cmfd_ != nullptr && cmfd_->solved() == false))) {
    prev_keff = keff_;
    keff_ = calc_keff(next_flux, flux_);
    rel_diff_keff = std::abs(keff_ - prev_keff) / keff_;
}
```

File: src/scarabee/_scarabee/include/moc/cmfd.hpp [hunk 22]
```diff
@@ -0,0 +1,213 @@
...
+
+  // Setter/Getter functions
+
+  double keff_tolerance() const { return keff_tol_; }
+  void set_keff_tolerance(double ktol);
+  double flux_tolerance() const { return flux_tol_; }
+  void set_flux_tolerance(double ftol);
+
+  void set_damping(double wd);
+  double get_damping() const { return damping_; }
```
[the remark this answers is not in the document]
Comment (gaiarj) [reply src/scarabee/_scarabee/include/moc/cmfd.hpp:106]: The function to do the Larsen correction is called larsen_correction so it will conflict. I can rename the actual larsen_correction function to something like apply_larsen_correction

File: src/scarabee/_scarabee/cmfd.cpp (hunk 18 above)
Comment (gaiarj) [reply src/scarabee/_scarabee/cmfd.cpp:719]: Is it faster to do EtDx*EtDx... and so on for each power?

File: src/scarabee/_scarabee/include/moc/cmfd.hpp (hunk 1 above)
Comment (gaiarj) [reply src/scarabee/_scarabee/include/moc/cmfd.hpp:210]: Seeing as it is always a member of the MOCDriver class, does the MOCDriver serialization need to be updated to reflect that at as well?

Comment (HunterBelanger) [reply src/scarabee/_scarabee/include/moc/cmfd.hpp:212]: Yes. So you need to make sure that the cmfd_ member of the MOCDriver gets serialized in the MOCDriver serializer, which requires that the CMFD class have its own serialization method.

File: src/scarabee/_scarabee/cmfd.cpp [hunk 23]
```diff
@@ -325,26 +460,758 @@ void CMFD::compute_homogenized_xs_and_flux(const MOCDriver& moc) {
...
+      }
+      if (std::isnan(ratio)) {
+        auto mssg = "CMFD flux update ratio is NaN";
+        spdlog::error(mssg);
+        throw ScarabeeException(mssg);
+      }
+      // Clamp CMFD flux ratio after a certain # of solves
+      if (cmfd_solves_ > unbounded_cmfd_solves_) {
+        std::clamp(ratio, 0.05, 20.0);
+      }
```
Comment (HunterBelanger) [inline src/scarabee/_scarabee/cmfd.cpp:1105]: Here, you are clamping the update ratio, but you never actually clamp / update the value in the update_ratios_ array.

File: src/scarabee/_scarabee/python/cmfd.cpp [hunk 24]
```diff
@@ -0,0 +1,183 @@
...
+
+  py::enum_<CMFDSurfaceCrossing::Type>(m,"CMFDSurfaceCrossingType")
+    .value("XN",CMFDSurfaceCrossing::Type::XN)
+    .value("XP",CMFDSurfaceCrossing::Type::XP)
+    .value("YN",CMFDSurfaceCrossing::Type::YN)
+    .value("YP",CMFDSurfaceCrossing::Type::YP)
+    .value("TR",CMFDSurfaceCrossing::Type::TR)
+    .value("BR",CMFDSurfaceCrossing::Type::BR)
+    .value("BL",CMFDSurfaceCrossing::Type::BL)
+    .value("TL",CMFDSurfaceCrossing::Type::TL);
```
Comment (HunterBelanger) [inline src/scarabee/_scarabee/python/cmfd.cpp:21]: These should all have description doc strings. Please check in other places to see how I add those for enum types.

File: src/scarabee/_scarabee/python/cmfd.cpp [hunk 25]
```diff
@@ -0,0 +1,183 @@
...
+  
+  py::class_<CMFDSurfaceCrossing>(m, "CMFDSurfaceCrossing")
+  .def_readonly("is_valid", &CMFDSurfaceCrossing::is_valid,
+  "Bool that indicates whether or not the segment end is on a valid CMFD surface")
+  .def_readonly("cell_index", &CMFDSurfaceCrossing::cell_index,
+  "Int representing the linear cell index")
+  .def_readonly("crossing", &CMFDSurfaceCrossing::crossing,
+  "Defines which corner / side of the cell the segment end passes through");
```
Comment (HunterBelanger) [inline src/scarabee/_scarabee/python/cmfd.cpp:29]:
```suggestion
  "Defines which side / corner of the cell the segment end passes through.");
```

File: src/scarabee/_scarabee/python/cmfd.cpp
Comment (HunterBelanger) [inline src/scarabee/_scarabee/python/cmfd.cpp:1]: As a general set of comments for this file, you want to make sure that all doc strings are proper sentences (which includes ending with a period).

Please make sure there is an blank line between each binding (this helps figure out where new bindings begin).

When you give the type for the doc strings, make sure you use Python types and not C++ types ! Something I see often here is you use `unsigned int`, however Python does not have an unsigned int type. You should just use int.

There are cases where you give the type hint twice (like the type, then in the description you repeat this). This isn't needed, and the redundant type can be removed from the description.

File: src/scarabee/_scarabee/python/cmfd.cpp [hunk 26]
```diff
@@ -0,0 +1,183 @@
...
+  .def("current",
+      py::overload_cast<const std::size_t, const std::size_t>(
+      &CMFD::current, py::const_),
+      "Returns the current on a CMFD cell boundary\n\n"
+      "Parameters\n"
+      "----------\n"
+      "G: unsigned int\n"
+      "   Int which gives the # of the flux energy group"
+      "surf: unsigned int\n"
+      "   Int which gives the CMFD surface index\n"
```
Comment (HunterBelanger) [inline src/scarabee/_scarabee/python/cmfd.cpp:147]: Here is a case where you don't need to say "Int which ...", because the int type is already specified on the line above.

**Review by HunterBelanger — COMMENTED** [review]
Here is a new round of comments. I have gone through everything except the tests. I still haven't looked at those. We can do those once the rest of the source is good.

File: src/scarabee/_scarabee/cmfd.cpp [hunk 27]
```diff
@@ -234,6 +306,27 @@ std::size_t CMFD::moc_to_cmfd_group(std::size_t g) const {
...
+                         const std::size_t g) const {
+  if (g > ng_) {
+    auto mssg = "Group index out of range.";
+    spdlog::error(mssg);
+    throw ScarabeeException(mssg);
+  }
+  if (i >= nx_) {
+    spdlog::error("Cell index {:d} is out of range", i);
+    throw ScarabeeException("Cell x index is out of range");
+  }
```
Comment (HunterBelanger) [inline src/scarabee/_scarabee/cmfd.cpp:319]: Please use identical error messages !!

File: src/scarabee/_scarabee/cmfd.cpp [hunk 28]
```diff
@@ -234,6 +306,27 @@ std::size_t CMFD::moc_to_cmfd_group(std::size_t g) const {
...
+  if (j >= nx_) {
+    spdlog::error("Cell index {:d} is out of range", j);
+    throw ScarabeeException("Cell y index is out of range");
+  }
```
Comment (HunterBelanger) [inline src/scarabee/_scarabee/cmfd.cpp:323]: Please use identical error messages !

File: src/scarabee/_scarabee/cmfd.cpp [hunk 29]
```diff
@@ -325,26 +476,767 @@ void CMFD::compute_homogenized_xs_and_flux(const MOCDriver& moc) {
...
+
+    // Loop over each FSR in CMFD cell i,j
+    for (std::size_t l = 0; l < tot_cells; l++) {
+      const auto& fsrs = fsrs_[l];
+      const std::size_t linear_indx = G * tot_cells + l;
+      const double& flx_ratio = update_ratios_(linear_indx);
+
+      // Update scalar flux in each MOC FSR
+      for (const auto f : fsrs) {
+        for (std::size_t lj = 0; lj < moc.num_spherical_harmonics(); lj++) {
```
[the remark this answers is not in the document]
Comment (gaiarj) [reply src/scarabee/_scarabee/cmfd.cpp:1150]: I don't remember seeing a difference either. We thought originally the non-convergence of anisotropic scattering past a certain tolerance was due to something related to this, but we know now that isotropic scattering has the same problem. So I think we can just do the 0th moment?

File: src/scarabee/_scarabee/cmfd.cpp [hunk 30]
```diff
@@ -325,26 +476,767 @@ void CMFD::compute_homogenized_xs_and_flux(const MOCDriver& moc) {
...
+        xt::view(track.exit_flux(), g, xt::all()) *=
+            update_ratios_(g_indx + exit_cell);
+      }
+    }
+  }
+
+  if (flux_update_warning) {
+    const double max_update = update_ratios_.maxCoeff();
+    spdlog::warn(
+        "At least one CMFD flux ratio greater than 20, max update ratio is: "
```
[the remark this answers is not in the document]
Comment (gaiarj) [reply src/scarabee/_scarabee/cmfd.cpp:1178]: right now it only checks if it is >20, I guess we should check if it is <0.05 as well? Although, if we implement the change to set the ratio = 0.0 when the flux is below a tolerance, that will cause a warning on every subsequent iteration

File: src/scarabee/_scarabee/include/moc/cmfd.hpp [hunk 31]
```diff
@@ -0,0 +1,291 @@
...
+  std::vector<XPlane> x_bounds_;
+  std::vector<YPlane> y_bounds_;
+  std::vector<std::size_t> moc_to_cmfd_group_map_;
+  std::vector<std::pair<std::size_t, std::size_t>> group_condensation_;
+  std::size_t nx_, ny_, ng_;
+  std::size_t nx_surfs_, ny_surfs_;
+
+  bool flux_limiting_ = true;
+  bool larsen_correction_ = false;
+  bool od_cmfd_ = false;
```
[the remark this answers is not in the document]
Comment (gaiarj) [reply src/scarabee/_scarabee/include/moc/cmfd.hpp:162]: C5G7 on my system: 
nothing - 17.1 s, 28 iterations
odCMFD - 21.5 s, 29 iterations
Larsen - 24.7 s, 30 iterations

Not an exhaustive test but it seems like using odCMFD shouldn't be too bad

File: src/scarabee/_scarabee/include/moc/moc_driver.hpp [hunk 32]
```diff
@@ -84,6 +105,17 @@ class MOCDriver {
               std::size_t lj = 0) const;
   double flux(std::size_t i, std::size_t g, std::size_t lj = 0) const;
 
+  void set_flux(std::size_t i, std::size_t g, double new_flx,
+                std::size_t lj = 0) {
+    flux_(g, i, lj) = new_flx;
+  }
```
[the remark this answers is not in the document]
Comment (gaiarj) [reply src/scarabee/_scarabee/include/moc/moc_driver.hpp:111]: I don't think you can exactly give friend access to a single function though? It is shared at the class level which is not what we want?

[the remark this answers is not in the document]
Comment (HunterBelanger) [reply src/scarabee/_scarabee/include/moc/moc_driver.hpp:111]: Ah no, you are right. I forget about that. I think just giving CMFD full friend access to MOCDriver is fine though in this case. Just then make the method private and still only use the method to change the flux.

File: src/scarabee/_scarabee/cmfd.cpp [hunk 33]
```diff
@@ -325,26 +477,788 @@ void CMFD::compute_homogenized_xs_and_flux(const MOCDriver& moc) {
...
+}
+
+void CMFD::set_larsen_correction(bool user_pref) {
+  if (od_cmfd_) {
+    auto mssg =
+        "odCMFD and Larsen Correction options are mutally exclusive. Disable "
+        "odCMFD with 'od_cmfd = False'.";
+    spdlog::error(mssg);
+    throw ScarabeeException(mssg);
+  }
```
Comment (HunterBelanger) [inline src/scarabee/_scarabee/cmfd.cpp:601]: This logic isn't quite correct. The user_pref could be false !
```suggestion
  if (od_cmfd_ && user_pref) {
    auto mssg =
        "odCMFD and Larsen Correction options are mutally exclusive. Disable "
        "odCMFD with 'od_cmfd = False'.";
    spdlog::error(mssg);
    throw ScarabeeException(mssg);
  }
```

File: tests/cmfd/test_cmfd_anisotropic.py
[the remark this answers is not in the document]
Comment (gaiarj) [reply tests/cmfd/test_cmfd_anisotropic.py:1]: How specific should I get with that? Should it just say that they were generated by creating cross sections for a reflected pin cell and using the cylindrical flux solver?

File: src/scarabee/_scarabee/cmfd.cpp (hunk 30 above)
[the remark this answers is not in the document]
Comment (gaiarj) [reply src/scarabee/_scarabee/cmfd.cpp:1198]: could a flux of exactly 0.0 cause problems with the invs_flx calculation, since it would divide by zero? That is the MOC homogenized flux though, so I think it should be ok because even if CMFD sets the flux to zero, the MOC might make it nonzero on the next sweep. I'm not sure

[the remark this answers is not in the document]
Comment (HunterBelanger) [reply src/scarabee/_scarabee/cmfd.cpp:1198]: I think it is very unlikely that we get a flux of exactly zero from the MOC solver, so I don't think it will be a problem in practice.

File: src/scarabee/_scarabee/python/cmfd.cpp (hunk 24 above)
Comment (HunterBelanger) [reply src/scarabee/_scarabee/python/cmfd.cpp]: That is because I am a lazy potato. The `DiffusionData` bindings have examples though ! Actually, you should directly copy some of those doc strings here.

File: tests/cmfd/test_cmfd_anisotropic.py
[the remark this answers is not in the document]
Comment (gaiarj) [reply tests/cmfd/test_cmfd_anisotropic.py:1]: It is essentially the pin_cell_spectrum.py example changed to include scattering moments, so its pretty long

File: src/scarabee/_scarabee/cmfd.cpp (hunk 29 above)
[the remark this answers is not in the document]
Comment (gaiarj) [reply src/scarabee/_scarabee/cmfd.cpp]: So, interestingly going back to only doing the 0th moment broke the UO2 assembly anisotropic example in test_cmfd_anisotropic - on the 2nd MOC iteration Dtr would become NaN. This is surprising to me because I thought we had tested this before and it worked... so I might've done something wrong

File: tests/cmfd/test_cmfd_anisotropic.py
[the remark this answers is not in the document]
Comment (HunterBelanger) [reply tests/cmfd/test_cmfd_anisotropic.py:1]: Ahh, okay. Just say that in some comments at the top of the file then.

**Comment by HunterBelanger** [comment]
@gaiarj I think this is in a state where this can be merged, if you have no objections ?
